opshin/type_inference.py

```diff
@@ -1010,6 +1010,29 @@ def visit_FunctionDef(self, node: FunctionDef) -> TypedFunctionDef:
     def visit_Module(self, node: Module) -> TypedModule:
         self.enter_scope()
         tm = copy(node)
+        
+        # First pass: extract all function signatures and add them to scope
+        # This enables mutual recursion by making all functions available before processing bodies
+        for stmt in node.body:
```

Comment:
this loop should occur in visit_sequence. notably, function definitions can also occur inside other functions, this needs to be taken into account (add a test case for this!)